### PR TITLE
docs: improve docstring of Rule

### PR DESCRIPTION
**Pull request by redeboer**
An attempt to make clearer that the type hints of the `Rule` derivative classes have an important dynamic role as well.

### CODE REVIEW

File: expertsystem/state/conservation_rules.py [hunk 1]
```diff
@@ -47,21 +47,24 @@ class Rule:
...
     overwritten.
 
+    .. warning::
+        Besides the rule logic itself, a `Rule` also has the responsibility of
+        stating its run conditions. These run conditions **must** be stated by
+        the type annotations of its `.__call__` method. The type annotations
+        therefore are not just there for *static* type checking: they also
+        carry more information about the rule that is extracted *dynamically*
+        by the `.propagation` module.
+
```
Comment (spflueger) [inline expertsystem/state/conservation_rules.py:57]: Is there a reason you moved this block above? I would keep things logically together so, this should go below the paragraph
> For additive quantum numbers, the decorator .....

Comment (redeboer) [reply expertsystem/state/conservation_rules.py:57]: Yeah I wasn't sure here. The `__annotations__` refer to the `__call__` method in the previous paragraphs, that's why I put it higher up

Comment (spflueger) [reply expertsystem/state/conservation_rules.py:57]: maybe move that additive quantum number paragraph to the end then?

Comment (redeboer) [reply expertsystem/state/conservation_rules.py]: Ok moved to the end

**Review by spflueger — APPROVED** [review]

**Pull request merged**